src/go/configmanager/config_manager.go

```diff
@@ -119,49 +122,94 @@ func NewConfigManager(mf *metadata.MetadataFetcher, opts options.ConfigGenerator
 		return nil, fmt.Errorf(`failed to set rollout strategy. It must be either "managed" or "fixed"`)
 	}
 
+	// when --non_gcp  is set, instance metadata server(imds) is not defined so
+	// accessToken is unavailable from imds and --service_account_key must be
+	// set to generate accessToken.
+	if mf == nil && opts.ServiceAccountKey == "" {
+		return nil, fmt.Errorf("If --non_gcp is specified, --service_account_key has to be specified.")
+	}
+
+	accessToken := func() (string, time.Duration, error) {
+		if opts.ServiceAccountKey != "" {
+			return util.GenerateAccessTokenFromFile(opts.ServiceAccountKey)
+		}
+		return mf.FetchAccessToken()
+	}
+
+	client, err := httpsClient(opts)
+	if err != nil {
+		return nil, fmt.Errorf("fail to init httpsClient: %v", err)
+	}
+
+	m.serviceConfigFetcher = sc.NewServiceConfigFetcher(client, opts.ServiceManagementURL,
+		m.serviceName, accessToken)
+
 	configId := ""
 	if rolloutStrategy == util.FixedRolloutStrategy {
-		// rollout strategy is fixed mode
 		configId = *ServiceConfigId
 		if configId == "" {
-			if checkMetadata && mf != nil {
-				configId, err = mf.FetchConfigId()
-				if configId == "" || err != nil {
-					return nil, fmt.Errorf("failed to read metadata with key endpoints-service-version from metadata server")
-				}
-			} else if !checkMetadata {
-				return nil, fmt.Errorf("service config id is not specified, required because metadata fetching is disabled")
-			} else if mf == nil {
+			if mf == nil {
 				return nil, fmt.Errorf("service config id is not specified, required on a non-gcp deployment")
 			}
-		}
 
-	}
+			if !checkMetadata {
+				return nil, fmt.Errorf("service config id is not specified, required because metadata fetching is disabled")
+			}
 
-	if m.serviceConfigFetcher, err = sc.NewServiceConfigFetcher(mf, opts, m.serviceName); err != nil {
-		return nil, fmt.Errorf(`failed to create https client to call ServiceManagement service, got error: %v`, err)
-	}
+			configId, err = mf.FetchConfigId()
+			if configId == "" || err != nil {
+				return nil, fmt.Errorf("failed to read metadata with key endpoints-service-version from metadata server")
+			}
+		}
+	} else if rolloutStrategy == util.ManagedRolloutStrategy {
+		m.RolloutIdDetector = sc.NewRolloutIdDetector(client, opts.ServiceControlURL, m.serviceName, accessToken)
+		latestRolloutId, err := m.RolloutIdDetector.FetchLatestRolloutId()
```

Comment:
At start,  we don't need to call RolloutIDDetector to get the latest rollout ID
we just call ServiceConfigFetcher to get the latest rollout info, that call doens't need any rollout_id.   Its returns will get the latest rollout id.  and the config_id to use, and remember this config_id as curr_config_id. 
Next get latest rollout may return the same config_id,  if so, skip it.

Reply:
I see.  you are fetching the rollout info by a specific rollout ID. That is why you need rollout ID first.

Hmm.  I like to keep the old way:  we always call for the latest rollout.  with this URL

"/v1/services/$serviceName/rollouts?filter=status=SUCCESS"

Empty report only to detect rollout change,  actually we don't need that rollout id.  

Once there is rollout change,
1) we always ask for the latest rollout,   from that to find out the config_id,  
2) if config_id is changed,  then fetch and apply it.